plumbing/transport/upload_pack.go

```diff
@@ -0,0 +1,246 @@
+}
+
+// UploadPack is a server command that serves the upload-pack service.
+func UploadPack(
+	ctx context.Context,
+	st storage.Storer,
+	r io.ReadCloser,
+	w io.WriteCloser,
+	opts *UploadPackOptions,
+) error {
+	if r == nil || w == nil {
+		return fmt.Errorf("nil reader or writer")
+	}
+
+	if opts == nil {
+		opts = &UploadPackOptions{}
+	}
+
+	switch version := ProtocolVersion(opts.GitProtocol); version {
+	case protocol.V2:
+		// TODO: support version 2
+	case protocol.V1:
+		if _, err := pktline.Writef(w, "version=%s\n", version.String()); err != nil {
+			return err
+		}
+		fallthrough
+	case protocol.V0:
+	default:
+		return fmt.Errorf("unknown protocol version %q", version)
+	}
+
+	if opts.AdvertiseRefs || !opts.StatelessRPC {
+		if err := AdvertiseReferences(ctx, st, w, UploadPackService, opts.StatelessRPC); err != nil {
+			return err
+		}
+	}
+
+	rd := bufio.NewReader(r)
+	if !opts.AdvertiseRefs {
+		l, _, err := pktline.PeekLine(rd)
+		if err != nil {
+			return err
+		}
+
+		// In case the client has nothing to send, it sends a flush packet to
+		// indicate that it is done sending data. In that case, we're done
+		// here.
+		if l == pktline.Flush {
+			return nil
+		}
+
+		// TODO: implement server negotiation algorithm
+		// Receive upload request
+
+		upreq := packp.NewUploadRequest()
+		if err := upreq.Decode(rd); err != nil {
+			return err
+		}
+
+		var (
+			wants = upreq.Wants
+			caps  = upreq.Capabilities
+		)
+
+		for {
+			// TODO: support multi_ack & multi_ack_detailed
+			_, p, err := pktline.PeekLine(rd)
+			if err != nil {
+				return err
+			}
+
+			if bytes.Equal(p, []byte("done\n")) {
+				// consume the "done" line
+				pktline.ReadLine(rd) // nolint: errcheck
+				break
+			}
+
+			// Consume line
+			pktline.ReadLine(rd) // nolint: errcheck
+		}
+
+		// Done with the request, now close the reader
+		// to indicate that we are done reading from it.
+		if err := r.Close(); err != nil {
+			return fmt.Errorf("closing reader: %s", err)
+		}
+
+		// TODO: support deepen, deepen-since, and deepen-not
+		var shupd packp.ShallowUpdate
+		if !upreq.Depth.IsZero() {
+			switch depth := upreq.Depth.(type) {
+			case packp.DepthCommits:
+				if err := getShallowCommits(st, upreq.Wants, int(depth), &shupd); err != nil {
+					return err
+				}
+			default:
+				return fmt.Errorf("unsupported depth type %T", upreq.Depth)
+			}
+
+			if err := shupd.Encode(w); err != nil {
+				return err
+			}
+		}
+
+		srvupd := packp.ServerResponse{}
+		if err := srvupd.Encode(w); err != nil {
+			return err
+		}
+
+		// Find common commits/objects
+		objs, err := objectsToUpload(st, wants, nil)
+		if err != nil {
+			return err
+		}
+
+		var writer io.Writer = w
+		if !caps.Supports(capability.NoProgress) {
+			if caps.Supports(capability.Sideband) {
+				writer = sideband.NewMuxer(sideband.Sideband, w)
+			}
+			if caps.Supports(capability.Sideband64k) {
+				writer = sideband.NewMuxer(sideband.Sideband64k, w)
+			}
```

Comment:
This is in inverse order. At some point we could look into refactor this, as the same logic is spread around 3 places (so far).

Reply:
Corrected the logic to be mutually exclusive. The server-side in `upload_pack.go` uses a *Muxer*, while the other two places in `fetch.go` and `push.go` client-side use a *Demuxer*.